platforms/ios/src/TangramMap/TGMarker.mm

```diff
@@ -7,244 +7,145 @@
 //
 
 #import "TGMarker.h"
+#import "TGMarker+Internal.h"
 #import "TGMapViewController.h"
 #import "TGMapViewController+Internal.h"
 #import "TGHelpers.h"
+#import "TGTypes.h"
 #import "tangram.h"
 
-enum class TGMarkerType {
-    point,
-    polygon,
-    polyline,
-    none,
-};
-
 @interface TGMarker () {
     Tangram::Map* tangramInstance;
-    Tangram::MarkerID identifier;
-    TGMarkerType type;
+    NSError* error;
 }
 
-@property (copy, nonatomic) NSString* stylingString;
-@property (copy, nonatomic) NSString* stylingPath;
-@property (assign, nonatomic) TGGeoPoint point;
-@property (strong, nonatomic) TGGeoPolyline* polyline;
-@property (strong, nonatomic) TGGeoPolygon* polygon;
-@property (assign, nonatomic) BOOL visible;
-@property (assign, nonatomic) NSInteger drawOrder;
-@property (strong, nonatomic) UIImage* icon;
-@property (weak, nonatomic) TGMapViewController* map;
-
-- (void)createNSError:(NSError **)error;
+- (void)createNSError;
+- (void)clearGeometry;
 
 @end
 
 @implementation TGMarker
 
-- (instancetype)init
+- (instancetype)initWithMap:(Tangram::Map*)map
 {
     self = [super init];
 
     if (self) {
-        type = TGMarkerType::none;
+        tangramInstance = map;
+        self.identifier = tangramInstance->markerAdd();
         self.visible = YES;
         self.drawOrder = 0;
     }
 
     return self;
 }
 
-- (instancetype)initWithMapView:(TGMapViewController *)mapView
+- (void)clearGeometry
 {
-    self = [super init];
-
-    if (self) {
-        type = TGMarkerType::none;
-
-        self.visible = YES;
-        self.drawOrder = 0;
-
-        [self map:mapView error:nil];
-    }
-
-    return self;
+    self.polyline = nil;
+    self.polygon = nil;
+    self.point = {NAN, NAN};
 }
 
-- (void)createNSError:(NSError **)error
+- (void)createNSError
 {
-    if (!error) {
-        return;
-    }
+    // TODO: Check for all Tangram internal tangram error and
```

Comment:
Since the informations given by the string itself and the marker API gives either success or failure, the error string here is quite simple. This string could be given more description details around the kind of error that actually happened if the core marker API evolves.

What is possible in this implementation is appending errors in the string as they come, to give more context and change the interface by adding the callee name:
```
[self createNSError:"setStylingString"];
```
But this wouldn't give you much more informations if you check errors after the API calls of interest.